File: launcher.cs
using System;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace TraceXLauncher
{
    public class MainForm : Form
    {
        private Button btnStart;
        private Button btnStop;
        private Button btnInstall;
        private Button btnOpenBrowser;
        private Button btnOpenDocs;
        private Button btnRunTests;
        private TextBox txtLog;
        private Label lblStatusBackend;
        private Label lblStatusFrontend;
        private Label lblTitle;
        private Label lblSubTitle;

        private Process backendProcess = null;
        private Process frontendProcess = null;
        private System.Windows.Forms.Timer statusTimer;

        private string projectRoot;
        private string pythonExe;
        private string venvPythonExe;
        private string npmExe = "npm.cmd";

        public MainForm()
        {
            projectRoot = AppDomain.CurrentDomain.BaseDirectory;
            venvPythonExe = Path.Combine(projectRoot, "backend", "venv", "Scripts", "python.exe");

            DetectPython();
            InitializeComponent();

            statusTimer = new System.Windows.Forms.Timer();
            statusTimer.Interval = 2500;
            statusTimer.Tick += StatusTimer_Tick;
            statusTimer.Start();

            this.FormClosing += MainForm_FormClosing;
        }

        private void DetectPython()
        {
            if (File.Exists(venvPythonExe))
            {
                pythonExe = venvPythonExe;
                return;
            }

            string userLocalPy = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "Programs", "Python", "Python311", "python.exe"
            );

            if (File.Exists(userLocalPy))
            {
                pythonExe = userLocalPy;
                return;
            }

            pythonExe = "python.exe";
        }

        private void InitializeComponent()
        {
            this.Text = "TRACE-X Pro - Launcher & Manager";
            this.Size = new Size(820, 620);
            this.StartPosition = FormStartPosition.CenterScreen;
            this.BackColor = Color.FromArgb(15, 23, 42); // Dark slate theme
            this.ForeColor = Color.FromArgb(241, 245, 249);
            this.Font = new Font("Segoe UI", 9.5f, FontStyle.Regular);

            // Header Title
            lblTitle = new Label();
            lblTitle.Text = "TRACE-X PRO";
            lblTitle.Font = new Font("Segoe UI", 18f, FontStyle.Bold);
            lblTitle.ForeColor = Color.FromArgb(56, 189, 248); // Cyan
            lblTitle.Location = new Point(24, 20);
            lblTitle.AutoSize = true;
            this.Controls.Add(lblTitle);

            lblSubTitle = new Label();
            lblSubTitle.Text = "AI-Powered Email Threat Detection & Google Workspace Ingestion Platform";
            lblSubTitle.Font = new Font("Segoe UI", 10f, FontStyle.Regular);
            lblSubTitle.ForeColor = Color.FromArgb(148, 163, 184);
            lblSubTitle.Location = new Point(26, 56);
            lblSubTitle.AutoSize = true;
            this.Controls.Add(lblSubTitle);

            // Status Panel
            Panel pnlStatus = new Panel();
            pnlStatus.BackColor = Color.FromArgb(30, 41, 59);
            pnlStatus.Location = new Point(24, 90);
            pnlStatus.Size = new Size(756, 44);
            this.Controls.Add(pnlStatus);

            lblStatusBackend = new Label();
            lblStatusBackend.Text = "Backend API: Stopped";
            lblStatusBackend.Location = new Point(16, 12);
            lblStatusBackend.AutoSize = true;
            lblStatusBackend.ForeColor = Color.FromArgb(248, 113, 113); // Red
            pnlStatus.Controls.Add(lblStatusBackend);

            lblStatusFrontend = new Label();
            lblStatusFrontend.Text = "Frontend UI: Stopped";
            lblStatusFrontend.Location = new Point(260, 12);
            lblStatusFrontend.AutoSize = true;
            lblStatusFrontend.ForeColor = Color.FromArgb(248, 113, 113);
            pnlStatus.Controls.Add(lblStatusFrontend);

            // Action Buttons
            btnStart = CreateStyledButton("Start TRACE-X", new Point(24, 150), Color.FromArgb(16, 185, 129)); // Green
            btnStart.Click += BtnStart_Click;
            this.Controls.Add(btnStart);

            btnStop = CreateStyledButton("Stop Services", new Point(154, 150), Color.FromArgb(239, 68, 68)); // Red
            btnStop.Click += BtnStop_Click;
            this.Controls.Add(btnStop);

            btnInstall = CreateStyledButton("Install / Repair", new Point(284, 150), Color.FromArgb(99, 102, 241)); // Indigo
            btnInstall.Click += BtnInstall_Click;
            this.Controls.Add(btnInstall);

            btnOpenBrowser = CreateStyledButton("Open Web UI", new Point(414, 150), Color.FromArgb(14, 165, 233)); // Sky
            btnOpenBrowser.Click += (s, e) => Process.Start("http://localhost:5173");
            this.Controls.Add(btnOpenBrowser);

            btnOpenDocs = CreateStyledButton("API Swagger", new Point(544, 150), Color.FromArgb(59, 130, 246));
            btnOpenDocs.Click += (s, e) => Process.Start("http://localhost:8000/docs");
            this.Controls.Add(btnOpenDocs);

            btnRunTests = CreateStyledButton("Run Tests", new Point(674, 150), Color.FromArgb(168, 85, 247)); // Purple
            btnRunTests.Click += BtnRunTests_Click;
            this.Controls.Add(btnRunTests);

            // Output Console Log
            txtLog = new TextBox();
            txtLog.Multiline = true;
            txtLog.ReadOnly = true;
            txtLog.ScrollBars = ScrollBars.Vertical;
            txtLog.BackColor = Color.FromArgb(10, 15, 26);
            txtLog.ForeColor = Color.FromArgb(226, 232, 240);
            txtLog.Font = new Font("Consolas", 9f);
            txtLog.Location = new Point(24, 206);
            txtLog.Size = new Size(756, 350);
            this.Controls.Add(txtLog);

            Log("TRACE-X Pro Launcher initialized.");
            Log("Project Directory: " + projectRoot);
            Log("Python Path: " + pythonExe);
            Log("Click 'Start TRACE-X' to run both Backend & Frontend.");
        }

        private Button CreateStyledButton(string text, Point location, Color color)
        {
            Button btn = new Button();
            btn.Text = text;
            btn.Location = location;
            btn.Size = new Size(120, 38);
            btn.FlatStyle = FlatStyle.Flat;
            btn.FlatAppearance.BorderSize = 0;
            btn.BackColor = color;
            btn.ForeColor = Color.White;
            btn.Font = new Font("Segoe UI", 9f, FontStyle.Bold);
            btn.Cursor = Cursors.Hand;
            return btn;
        }

        private void Log(string message)
        {
            if (txtLog.InvokeRequired)
            {
                txtLog.Invoke(new Action<string>(Log), message);
                return;
            }
            string time = DateTime.Now.ToString("HH:mm:ss");
            txtLog.AppendText("[" + time + "] " + message + Environment.NewLine);
        }

        private async void BtnStart_Click(object sender, EventArgs e)
        {
            Log("Starting TRACE-X Pro services...");

            // 1. Ensure backend virtualenv exists
            if (!File.Exists(venvPythonExe))
            {
                Log("Virtual environment not found. Running auto-install first...");
                await RunInstallRoutine();
            }

            // 2. Start Backend API
            StartBackend();

            // 3. Start Frontend UI
            StartFrontend();

            // 4. Wait 3 seconds and open browser
            await Task.Delay(3500);
            try
            {
                Process.Start("http://localhost:5173");
                Log("Browser opened to http://localhost:5173");
            }
            catch { }
        }

        private void StartBackend()
        {
            if (backendProcess != null && !backendProcess.HasExited)
            {
                Log("Backend is already running.");
                return;
            }

            try
            {
                string backendDir = Path.Combine(projectRoot, "backend");
                string py = File.Exists(venvPythonExe) ? venvPythonExe : pythonExe;

                ProcessStartInfo psi = new ProcessStartInfo();
                psi.FileName = py;
                psi.Arguments = "-m uvicorn app.main:app --host 127.0.0.1 --port 8000";
                psi.WorkingDirectory = backendDir;
                psi.UseShellExecute = false;
                psi.RedirectStandardOutput = true;
                psi.RedirectStandardError = true;
                psi.CreateNoWindow = true;

                backendProcess = new Process();
                backendProcess.StartInfo = psi;
                backendProcess.OutputDataReceived += (s, e) => { if (e.Data != null) Log("[API] " + e.Data); };
                backendProcess.ErrorDataReceived += (s, e) => { if (e.Data != null) Log("[API] " + e.Data); };

                backendProcess.Start();
                backendProcess.BeginOutputReadLine();
                backendProcess.BeginErrorReadLine();

                Log("FastAPI Backend started on http://127.0.0.1:8000 (PID: " + backendProcess.Id + ")");
            }
            catch (Exception ex)
            {
                Log("ERROR starting backend: " + ex.Message);
            }
        }

        private void StartFrontend()
        {
            if (frontendProcess != null && !frontendProcess.HasExited)
            {
                Log("Frontend is already running.");
                return;
            }

            try
            {
                string frontendDir = Path.Combine(projectRoot, "frontend");

                ProcessStartInfo psi = new ProcessStartInfo();
                psi.FileName = "cmd.exe";
                psi.Arguments = "/c npm run dev";
                psi.WorkingDirectory = frontendDir;
                psi.UseShellExecute = false;
                psi.RedirectStandardOutput = true;
                psi.RedirectStandardError = true;
                psi.CreateNoWindow = true;

                frontendProcess = new Process();
                frontendProcess.StartInfo = psi;
                frontendProcess.OutputDataReceived += (s, e) => { if (e.Data != null) Log("[UI] " + e.Data); };
                frontendProcess.ErrorDataReceived += (s, e) => { if (e.Data != null) Log("[UI] " + e.Data); };

                frontendProcess.Start();
                frontendProcess.BeginOutputReadLine();
                frontendProcess.BeginErrorReadLine();

                Log("Vite Frontend started on http://localhost:5173 (PID: " + frontendProcess.Id + ")");
            }
            catch (Exception ex)
            {
                Log("ERROR starting frontend: " + ex.Message);
            }
        }

        private void BtnStop_Click(object sender, EventArgs e)
        {
            StopServices();
        }

        private void StopServices()
        {
            Log("Stopping services...");

            if (backendProcess != null && !backendProcess.HasExited)
            {
                try
                {
                    backendProcess.Kill();
                    Log("Backend process stopped.");
                }
                catch { }
                backendProcess = null;
            }

            if (frontendProcess != null && !frontendProcess.HasExited)
            {
                try
                {
                    // Kill process tree
                    Process.Start(new ProcessStartInfo
                    {
                        FileName = "taskkill.exe",
                        Arguments = "/PID " + frontendProcess.Id + " /T /F",
                        CreateNoWindow = true,
                        UseShellExecute = false
                    });
                    Log("Frontend process stopped.");
                }
                catch { }
                frontendProcess = null;
            }

            lblStatusBackend.Text = "Backend API: Stopped";
            lblStatusBackend.ForeColor = Color.FromArgb(248, 113, 113);
            lblStatusFrontend.Text = "Frontend UI: Stopped";
            lblStatusFrontend.ForeColor = Color.FromArgb(248, 113, 113);
        }

        private async void BtnInstall_Click(object sender, EventArgs e)
        {
            await RunInstallRoutine();
        }

        private async Task RunInstallRoutine()
        {
            Log("==== Starting Automated Installation & Environment Setup ====");

            await Task.Run(() =>
            {
                try
                {
                    string backendDir = Path.Combine(projectRoot, "backend");
                    string venvDir = Path.Combine(backendDir, "venv");
                    string reqFile = Path.Combine(backendDir, "requirements.txt");

                    // 1. Create virtualenv
                    if (!File.Exists(venvPythonExe))
                    {
                        Log("Creating Python virtual environment in backend\\venv...");
                        Process p = Process.Start(new ProcessStartInfo
                        {
                            FileName = pythonExe,
                            Arguments = "-m venv \"" + venvDir + "\"",
                            WorkingDirectory = backendDir,
                            CreateNoWindow = true,
                            UseShellExecute = false
                        });
                        p.WaitForExit();
                        Log("Virtual environment created successfully.");
                    }

                    // 2. Install requirements
                    Log("Installing backend dependencies from requirements.txt...");
                    Process pip = Process.Start(new ProcessStartInfo
                    {
                        FileName = Path.Combine(venvDir, "Scripts", "pip.exe"),
                        Arguments = "install -r \"" + reqFile + "\"",
                        WorkingDirectory = backendDir,
                        CreateNoWindow = true,
                        UseShellExecute = false
                    });
                    pip.WaitForExit();
                    Log("Backend requirements installation completed.");

                    // 3. Setup .env file
                    string envFile = Path.Combine(backendDir, ".env");
                    string envExample = Path.Combine(backendDir, ".env.example");
                    if (!File.Exists(envFile) && File.Exists(envExample))
                    {
                        File.Copy(envExample, envFile);
                        Log("Created backend\\.env configuration from template.");
                    }

                    Log("==== Installation Complete! Ready to Launch. ====");
                }
                catch (Exception ex)
                {
                    Log("Installation error: " + ex.Message);
                }
            });
        }

        private async void BtnRunTests_Click(object sender, EventArgs e)
        {
            Log("Running pytest test suite across TRACE-X Pro...");

            await Task.Run(() =>
            {
                try
                {
                    string backendDir = Path.Combine(projectRoot, "backend");
                    string py = File.Exists(venvPythonExe) ? venvPythonExe : pythonExe;

                    ProcessStartInfo psi = new ProcessStartInfo();
                    psi.FileName = py;
                    psi.Arguments = "-m pytest tests -v";
                    psi.WorkingDirectory = backendDir;
                    psi.UseShellExecute = false;
                    psi.RedirectStandardOutput = true;
                    psi.RedirectStandardError = true;
                    psi.CreateNoWindow = true;

                    Process p = new Process();
                    p.StartInfo = psi;
                    p.OutputDataReceived += (s, args) => { if (args.Data != null) Log(args.Data); };
                    p.ErrorDataReceived += (s, args) => { if (args.Data != null) Log(args.Data); };

                    p.Start();
                    p.BeginOutputReadLine();
                    p.BeginErrorReadLine();
                    p.WaitForExit();

                    Log("Test suite completed with exit code: " + p.ExitCode);
                }
                catch (Exception ex)
                {
                    Log("Error running tests: " + ex.Message);
                }
            });
        }

        private async void StatusTimer_Tick(object sender, EventArgs e)
        {
            using (HttpClient client = new HttpClient())
            {
                client.Timeout = TimeSpan.FromMilliseconds(1200);

                // Check Backend
                try
                {
                    HttpResponseMessage resp = await client.GetAsync("http://127.0.0.1:8000/health");
                    if (resp.IsSuccessStatusCode)
                    {
                        lblStatusBackend.Text = "Backend API: ONLINE (Port 8000)";
                        lblStatusBackend.ForeColor = Color.FromArgb(52, 211, 153); // Green
                    }
                    else
                    {
                        lblStatusBackend.Text = "Backend API: Degraded";
                        lblStatusBackend.ForeColor = Color.FromArgb(251, 191, 36); // Yellow
                    }
                }
                catch
                {
                    lblStatusBackend.Text = "Backend API: Stopped";
                    lblStatusBackend.ForeColor = Color.FromArgb(248, 113, 113);
                }

                // Check Frontend
                try
                {
                    HttpResponseMessage resp = await client.GetAsync("http://localhost:5173");
                    if (resp.IsSuccessStatusCode)
                    {
                        lblStatusFrontend.Text = "Frontend UI: ONLINE (Port 5173)";
                        lblStatusFrontend.ForeColor = Color.FromArgb(52, 211, 153);
                    }
                    else
                    {
                        lblStatusFrontend.Text = "Frontend UI: Degraded";
                        lblStatusFrontend.ForeColor = Color.FromArgb(251, 191, 36);
                    }
                }
                catch
                {
                    lblStatusFrontend.Text = "Frontend UI: Stopped";
                    lblStatusFrontend.ForeColor = Color.FromArgb(248, 113, 113);
                }
            }
        }

        private void MainForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            StopServices();
        }

        [STAThread]
        public static void Main()
        {
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new MainForm());
        }
    }
}

File: launcher.py
"""
TRACE-X Pro - Antivirus-Safe Python Desktop Launcher & Control Center
Runs directly via Python's officially signed pythonw.exe without triggering antivirus false positives.
"""
import os
import sys
import subprocess
import threading
import time
import webbrowser
import urllib.request
import tkinter as tk
from tkinter import ttk, messagebox, scrolledtext

PROJECT_ROOT = os.path.dirname(os.path.abspath(__file__))
BACKEND_DIR = os.path.join(PROJECT_ROOT, "backend")
FRONTEND_DIR = os.path.join(PROJECT_ROOT, "frontend")
VENV_PYTHON = os.path.join(BACKEND_DIR, "venv", "Scripts", "python.exe")
PYTHON_EXE = VENV_PYTHON if os.path.exists(VENV_PYTHON) else sys.executable


class TraceXLauncherApp:
    def __init__(self, root):
        self.root = root
        self.root.title("TRACE-X PRO - Control Center")
        self.root.geometry("860x650")
        self.root.configure(bg="#f8fafc")

        self.backend_proc = None
        self.frontend_proc = None
        self._running = True
        self.logo_img = None

        self._create_widgets()
        self._start_status_polling()

        self.root.protocol("WM_DELETE_WINDOW", self.on_close)

    def _create_widgets(self):
        # Header frame
        header_frame = tk.Frame(self.root, bg="#ffffff", padx=24, pady=16, highlightthickness=1, highlightbackground="#e2e8f0")
        header_frame.pack(fill="x", padx=20, pady=(16, 8))

        # Logo & Title Container
        title_container = tk.Frame(header_frame, bg="#ffffff")
        title_container.pack(side="left", fill="both")

        # Try loading logo image
        logo_path = os.path.join(PROJECT_ROOT, "logo.png")
        if os.path.exists(logo_path):
            try:
                from PIL import Image, ImageTk
                pil_img = Image.open(logo_path)
                aspect = pil_img.width / pil_img.height
                target_height = 42
                target_width = int(target_height * aspect)
                pil_img = pil_img.resize((target_width, target_height), Image.Resampling.LANCZOS)
                self.logo_img = ImageTk.PhotoImage(pil_img)
                logo_lbl = tk.Label(title_container, image=self.logo_img, bg="#ffffff")
                logo_lbl.pack(side="left", padx=(0, 14))
            except Exception:
                pass

        text_container = tk.Frame(title_container, bg="#ffffff")
        text_container.pack(side="left", fill="both")

        title_lbl = tk.Label(
            text_container,
            text="TRACE-X PRO",
            font=("Segoe UI", 18, "bold"),
            fg="#dc2626",
            bg="#ffffff"
        )
        title_lbl.pack(anchor="w")

        subtitle_lbl = tk.Label(
            text_container,
            text="AI-Powered Email Threat Detection & Real-Time Ingestion Platform",
            font=("Segoe UI", 9),
            fg="#64748b",
            bg="#ffffff"
        )
        subtitle_lbl.pack(anchor="w", pady=(1, 0))

        # Status Bar Panel
        status_panel = tk.Frame(self.root, bg="#ffffff", padx=20, pady=12, highlightthickness=1, highlightbackground="#e2e8f0")
        status_panel.pack(fill="x", padx=20, pady=8)

        self.backend_status_lbl = tk.Label(
            status_panel,
            text="Backend API: Stopped",
            font=("Segoe UI", 9, "bold"),
            fg="#ef4444",
            bg="#ffffff"
        )
        self.backend_status_lbl.pack(side="left", padx=(0, 40))

        self.frontend_status_lbl = tk.Label(
            status_panel,
            text="Frontend UI: Stopped",
            font=("Segoe UI", 9, "bold"),
            fg="#ef4444",
            bg="#ffffff"
        )
        self.frontend_status_lbl.pack(side="left")

        # Action Buttons Frame
        btn_frame = tk.Frame(self.root, bg="#f8fafc")
        btn_frame.pack(fill="x", padx=20, pady=8)

        # Style helpers
        btn_config = {
            "font": ("Segoe UI", 9, "bold"),
            "relief": "flat",
            "cursor": "hand2",
            "padx": 14,
            "pady": 8,
            "fg": "#ffffff"
        }

        self.btn_start = tk.Button(
            btn_frame,
            text="▶ Start TRACE-X",
            bg="#dc2626",
            activebackground="#b91c1c",
            command=self.start_services,
            **btn_config
        )
        self.btn_start.pack(side="left", padx=(0, 8))

        self.btn_stop = tk.Button(
            btn_frame,
            text="⏹ Stop Services",
            bg="#64748b",
            activebackground="#475569",
            command=self.stop_services,
            **btn_config
        )
        self.btn_stop.pack(side="left", padx=(0, 8))

        self.btn_open_ui = tk.Button(
            btn_frame,
            text="🌐 Open Web UI",
            bg="#0f172a",
            activebackground="#1e293b",
            command=lambda: webbrowser.open("http://localhost:5173"),
            **btn_config
        )
        self.btn_open_ui.pack(side="left", padx=(0, 8))

        self.btn_docs = tk.Button(
            btn_frame,
            text="📖 API Swagger",
            bg="#ffffff",
            fg="#334155",
            highlightthickness=1,
            highlightbackground="#cbd5e1",
            activebackground="#f1f5f9",
            command=lambda: webbrowser.open("http://localhost:8000/docs"),
            font=("Segoe UI", 9, "bold"),
            relief="flat",
            cursor="hand2",
            padx=14,
            pady=8
        )
        self.btn_docs.pack(side="left", padx=(0, 8))

        self.btn_tests = tk.Button(
            btn_frame,
            text="🧪 Run Tests",
            bg="#ffffff",
            fg="#334155",
            highlightthickness=1,
            highlightbackground="#cbd5e1",
            activebackground="#f1f5f9",
            command=self.run_tests_threaded,
            font=("Segoe UI", 9, "bold"),
            relief="flat",
            cursor="hand2",
            padx=14,
            pady=8
        )
        self.btn_tests.pack(side="left")

        # Console Log Window
        log_label = tk.Label(
            self.root,
            text="Live Activity & Diagnostics Log:",
            font=("Segoe UI", 9, "bold"),
            fg="#475569",
            bg="#f8fafc"
        )
        log_label.pack(anchor="w", padx=20, pady=(10, 4))

        self.log_box = scrolledtext.ScrolledText(
            self.root,
            bg="#ffffff",
            fg="#0f172a",
            insertbackground="#dc2626",
            font=("Consolas", 9),
            relief="flat",
            highlightthickness=1,
            highlightbackground="#cbd5e1",
            wrap="word"
        )
        self.log_box.pack(fill="both", expand=True, padx=20, pady=(0, 16))

        self.log("TRACE-X Pro Control Center initialized.")
        self.log(f"Project directory: {PROJECT_ROOT}")
        self.log(f"Python interpreter: {PYTHON_EXE}")
        self.log("Click '▶ Start TRACE-X' to launch Backend API and Frontend UI.")

    def log(self, text):
        timestamp = time.strftime("%H:%M:%S")
        self.log_box.insert(tk.END, f"[{timestamp}] {text}\n")
        self.log_box.see(tk.END)

    def start_services(self):
        self.log("Starting TRACE-X services...")

        # 1. Start Backend API
        if not self.backend_proc or self.backend_proc.poll() is not None:
            cmd = [PYTHON_EXE, "-m", "uvicorn", "app.main:app", "--host", "127.0.0.1", "--port", "8000"]
            try:
                self.backend_proc = subprocess.Popen(
                    cmd,
                    cwd=BACKEND_DIR,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.STDOUT,
                    text=True,
                    bufsize=1,
                    creationflags=subprocess.CREATE_NO_WINDOW if os.name == "nt" else 0
                )
                threading.Thread(target=self._stream_proc_output, args=(self.backend_proc, "API"), daemon=True).start()
                self.log(f"FastAPI Backend process launched (PID: {self.backend_proc.pid})")
            except Exception as e:
                self.log(f"ERROR launching backend: {e}")

        # 2. Start Frontend UI
        if not self.frontend_proc or self.frontend_proc.poll() is not None:
            cmd = "npm run dev"
            try:
                self.frontend_proc = subprocess.Popen(
                    cmd,
                    cwd=FRONTEND_DIR,
                    shell=True,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.STDOUT,
                    text=True,
                    bufsize=1,
                    creationflags=subprocess.CREATE_NO_WINDOW if os.name == "nt" else 0
                )
                threading.Thread(target=self._stream_proc_output, args=(self.frontend_proc, "UI"), daemon=True).start()
                self.log(f"Vite Frontend process launched (PID: {self.frontend_proc.pid})")
            except Exception as e:
                self.log(f"ERROR launching frontend: {e}")

        # Open browser in 3.5 seconds
        threading.Thread(target=self._delayed_browser_open, daemon=True).start()

    def _stream_proc_output(self, proc, tag):
        try:
            for line in iter(proc.stdout.readline, ""):
                if not self._running:
                    break
                stripped = line.strip()
                if stripped:
                    self.root.after(0, self.log, f"[{tag}] {stripped}")
        except Exception:
            pass

    def _delayed_browser_open(self):
        time.sleep(3.5)
        webbrowser.open("http://localhost:5173")
        self.root.after(0, self.log, "Browser opened to http://localhost:5173")

    def stop_services(self):
        self.log("Stopping all TRACE-X services...")
        if self.backend_proc and self.backend_proc.poll() is None:
            try:
                self.backend_proc.terminate()
                self.backend_proc.kill()
                self.log("Backend API stopped.")
            except Exception:
                pass
            self.backend_proc = None

        if self.frontend_proc and self.frontend_proc.poll() is None:
            try:
                if os.name == "nt":
                    subprocess.run(["taskkill", "/PID", str(self.frontend_proc.pid), "/T", "/F"], capture_output=True)
                else:
                    self.frontend_proc.kill()
                self.log("Frontend UI stopped.")
            except Exception:
                pass
            self.frontend_proc = None

        self.backend_status_lbl.configure(text="Backend API: Stopped", fg="#f87171")
        self.frontend_status_lbl.configure(text="Frontend UI: Stopped", fg="#f87171")

    def run_tests_threaded(self):
        self.log("Starting pytest test suite in background...")
        threading.Thread(target=self._execute_tests, daemon=True).start()

    def _execute_tests(self):
        try:
            cmd = [PYTHON_EXE, "-m", "pytest", "tests", "-v"]
            p = subprocess.Popen(
                cmd,
                cwd=BACKEND_DIR,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
                creationflags=subprocess.CREATE_NO_WINDOW if os.name == "nt" else 0
            )
            for line in iter(p.stdout.readline, ""):
                stripped = line.strip()
                if stripped:
                    self.root.after(0, self.log, f"[TEST] {stripped}")
            p.wait()
            self.root.after(0, self.log, f"Test suite finished with code: {p.returncode}")
        except Exception as e:
            self.root.after(0, self.log, f"Error running tests: {e}")

    def _start_status_polling(self):
        def poll():
            while self._running:
                # Backend check
                backend_online = False
                try:
                    req = urllib.request.Request("http://127.0.0.1:8000/health")
                    with urllib.request.urlopen(req, timeout=1.2) as response:
                        if response.status == 200:
                            backend_online = True
                except Exception:
                    backend_online = False

                # Frontend check
                frontend_online = False
                try:
                    req = urllib.request.Request("http://localhost:5173")
                    with urllib.request.urlopen(req, timeout=1.2) as response:
                        if response.status == 200:
                            frontend_online = True
                except Exception:
                    frontend_online = False

                def update_labels():
                    if backend_online:
                        self.backend_status_lbl.configure(text="Backend API: ONLINE (Port 8000)", fg="#34d399")
                    else:
                        self.backend_status_lbl.configure(text="Backend API: Stopped", fg="#f87171")

                    if frontend_online:
                        self.frontend_status_lbl.configure(text="Frontend UI: ONLINE (Port 5173)", fg="#34d399")
                    else:
                        self.frontend_status_lbl.configure(text="Frontend UI: Stopped", fg="#f87171")

                try:
                    self.root.after(0, update_labels)
                except Exception:
                    break
                time.sleep(2.5)

        threading.Thread(target=poll, daemon=True).start()

    def on_close(self):
        self._running = False
        self.stop_services()
        self.root.destroy()


if __name__ == "__main__":
    root = tk.Tk()
    app = TraceXLauncherApp(root)
    root.mainloop()
